ee/tools/mdm/certs.go

```diff
@@ -0,0 +1,257 @@
+	if err != nil {
+		log.Fatalf("failed to load vendor private key: %s", err.Error())
+	}
+
+	// Decode CSR input
+	// We accept the CSR via environment variable to mitigate against command injection attacks in
+	// the fleetdm.com website code that will call this with untrusted user input.
+	csrBase64 := os.Getenv(csrEnvName)
+	if csrBase64 == "" {
+		log.Fatalf("CSR must be set in %s", csrEnvName)
+	}
+	csr, err := base64.StdEncoding.DecodeString(string(csrBase64))
+	if err != nil {
+		log.Fatalf("base64 decode csr: %s", err.Error())
+	}
+	certReq, err := decodeCSR(csr)
+	if err != nil {
+		log.Fatalf("decode pem: %s", err.Error())
+	}
+
+	// Get email from CSR
+	email, err := getEmail(certReq)
+	if err != nil {
+		log.Fatalf("get email: %s", err.Error())
+	}
+
+	// Tie it all together
+	req, err := createPushCertificateRequest(vendorKey, vendorCert, certReq)
+	if err != nil {
+		log.Fatalf("create request: %s", err.Error())
+	}
+	encodedReq, err := req.Encode()
+	if err != nil {
+		log.Fatalf("encode csr: %v", err)
+	}
+
+	// Write output as JSON
+	out := struct {
+		Email   string `json:"email"`
+		Request string `json:"request"`
+	}{
+		Email:   email,
+		Request: string(encodedReq),
+	}
+	outJSON, err := json.Marshal(out)
+	if err != nil {
+		log.Fatalf("encode request JSON: %s", err.Error())
+	}
+	fmt.Println(string(outJSON))
+}
+
+func getEmail(req *x509.CertificateRequest) (string, error) {
+	for _, name := range req.Subject.Names {
+		if slices.Equal(name.Type, emailAddressOID) {
+			str, ok := name.Value.(string)
+			if !ok {
+				return "", errors.New("email subject is not string value")
+			}
+			return str, nil
+		}
+	}
+	return "", errors.New("missing email subject")
+}
+
+// Below functions copied with modifications from MicroMDM (MIT license):
+// https://github.com/micromdm/micromdm/blob/e0943fdcd7ea2c6f79edf3e3496e1218e968ba2a/pkg/crypto/mdmcertutil/certutil.go
+
+// createPushCertificateRequest creates a request structure required by identity.apple.com.
+// It requires a "MDM CSR" certificate (the vendor certificate), a push CSR (the customer specific CSR),
+// and the vendor private key.
+func createPushCertificateRequest(vendorKey *rsa.PrivateKey, vendorCert *x509.Certificate, csr *x509.CertificateRequest) (*mdmcertutil.PushCertificateRequest, error) {
+	// csr signature
+	signature, err := signPushCSR(csr.Raw, vendorKey)
+	if err != nil {
+		return nil, fmt.Errorf("sign push CSR with private key: %w", err)
+	}
+
+	// vendor cert
+	mdmPEM := pemCert(vendorCert.Raw)
+
+	// wwdr cert
+	wwdrCertBytes, err := loadCertfromHTTP(wwdrIntermediaryURL)
+	if err != nil {
+		return nil, fmt.Errorf("load WWDR certificate from %s: %w", wwdrIntermediaryURL, err)
+	}
+	wwdrPEM := pemCert(wwdrCertBytes)
+
+	// apple root certificate
+	rootCertBytes, err := loadCertfromHTTP(appleRootCAURL)
+	if err != nil {
+		return nil, fmt.Errorf("load root certificate from %s: %w", appleRootCAURL, err)
+	}
+	rootPEM := pemCert(rootCertBytes)
```

Comment:
Perhaps we should just store this whole cert chain as a constant? This is copied from the MicroMDM code, but I don't anticipate the cert chain would change very often for our purposes.

Reply:
What would the advantage be of putting it in as a constant? 1 fewer request to apple? Or are we worried about the certificate changing without our knowledge?